File: di-ipv-cimit-stub/lambdas/stub-management/src/test/java/uk/gov/di/ipv/core/stubmanagement/StubManagementHandlerTest.java
package uk.gov.di.ipv.core.stubmanagement;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import uk.gov.di.ipv.core.library.model.UserMitigationRequest;
import uk.gov.di.ipv.core.library.persistence.items.CimitStubItem;
import uk.gov.di.ipv.core.library.service.CimitStubItemService;
import uk.gov.di.ipv.core.library.service.PendingMitigationService;
import uk.gov.di.ipv.core.stubmanagement.exceptions.BadRequestException;
import uk.gov.di.ipv.core.stubmanagement.exceptions.DataNotFoundException;
import uk.gov.di.ipv.core.stubmanagement.model.UserCisRequest;
import uk.gov.di.ipv.core.stubmanagement.service.UserService;

import java.io.IOException;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class StubManagementHandlerTest {

    @Mock private UserService mockUserService;
    @Mock private PendingMitigationService mockPendingMitigationService;
    @Mock private CimitStubItemService mockCimitStubItemService;
    @InjectMocks private StubManagementHandler stubManagementHandler;

    private final ObjectMapper objectMapper = new ObjectMapper();

    @Test
    public void shouldAddUserCisSuccessWhenValidCiRequestList() throws IOException {
        List<UserCisRequest> userCisRequests =
                List.of(
                        UserCisRequest.builder()
                                .code("code1")
                                .issuanceDate("2023-07-25T10:00:00Z")
                                .mitigations(List.of("V01", "V03"))
                                .build(),
                        UserCisRequest.builder()
                                .code("code2")
                                .issuanceDate("2023-07-25T10:00:00Z")
                                .mitigations(Collections.emptyList())
                                .build());

        APIGatewayProxyRequestEvent event =
                createTestEvent("POST", "/user/123/cis", userCisRequests);
        doNothing().when(mockUserService).addUserCis(anyString(), anyList());

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(200, response.getStatusCode());
        assertTrue(response.getBody().contains("success"));
        verify(mockUserService, times(1)).addUserCis(anyString(), anyList());
    }

    @Test
    public void shouldAUpdateUserCisSuccessWhenValidCiRequestList() throws IOException {
        UserCisRequest userCisRequest =
                UserCisRequest.builder()
                        .code("code1")
                        .issuanceDate("2023-07-25T10:00:00Z")
                        .mitigations(List.of("V01", "V03"))
                        .build();
        APIGatewayProxyRequestEvent event =
                createTestEvent("PUT", "/user/123/cis", Collections.singletonList(userCisRequest));
        doNothing().when(mockUserService).updateUserCis(anyString(), anyList());

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(200, response.getStatusCode());
        assertTrue(response.getBody().contains("success"));
        verify(mockUserService)
                .updateUserCis(eq("123"), eq(Collections.singletonList(userCisRequest)));
    }

    @ParameterizedTest
    @ValueSource(strings = {"POST", "PUT"})
    void shouldAddPendingMitigationWhenValidMitigationRequest(String method) throws IOException {
        when(mockCimitStubItemService.getCiForUserId("123", "456")).thenReturn(new CimitStubItem());

        UserMitigationRequest userMitigationRequest =
                UserMitigationRequest.builder().mitigations(List.of("V01")).build();

        APIGatewayProxyRequestEvent event =
                createTestEvent(method, "/user/123/mitigations/456", userMitigationRequest);

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(200, response.getStatusCode());
        assertTrue(response.getBody().contains("success"));
        verify(mockPendingMitigationService)
                .persistPendingMitigation(userMitigationRequest, "456", method);
    }

    @ParameterizedTest
    @ValueSource(strings = {"POST", "PUT"})
    void shouldReturn404IfCiItemNotFoundForMitigationRequest(String method) throws Exception {
        when(mockCimitStubItemService.getCiForUserId("123", "456")).thenReturn(null);

        UserMitigationRequest userMitigationRequest =
                UserMitigationRequest.builder().mitigations(List.of("V01")).build();

        APIGatewayProxyRequestEvent event =
                createTestEvent(method, "/user/123/mitigations/456", userMitigationRequest);

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(404, response.getStatusCode());
        assertTrue(response.getBody().contains("not found"));
        verify(mockPendingMitigationService, times(0))
                .persistPendingMitigation(userMitigationRequest, "456", method);
    }

    @Test
    void cisPatternShouldHandleDefaultUserIdFormat() throws Exception {
        String urlEncodedUserId = "urn%3Auuid%3Ac08630f8-330e-43f8-a782-21432a197fc5";
        List<UserCisRequest> userCisRequests =
                List.of(
                        UserCisRequest.builder()
                                .code("code1")
                                .issuanceDate("2023-07-25T10:00:00Z")
                                .mitigations(List.of("V01", "V03"))
                                .build());

        APIGatewayProxyRequestEvent event =
                createTestEvent(
                        "POST", String.format("/user/%s/cis", urlEncodedUserId), userCisRequests);
        doNothing().when(mockUserService).addUserCis(anyString(), anyList());

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(200, response.getStatusCode());
        assertTrue(response.getBody().contains("success"));
        verify(mockUserService, times(1)).addUserCis(anyString(), anyList());
    }

    @Test
    void mitigationsPatternShouldHandleDefaultUserIdFormat() throws Exception {
        when(mockCimitStubItemService.getCiForUserId("123", "456")).thenReturn(new CimitStubItem());

        String urlEncodedUserId = "urn%3Auuid%3Ac08630f8-330e-43f8-a782-21432a197fc5";
        UserMitigationRequest userMitigationRequest =
                UserMitigationRequest.builder().mitigations(List.of("V01")).build();

        APIGatewayProxyRequestEvent event =
                createTestEvent(
                        "POST",
                        String.format("/user/%s/mitigations/456", urlEncodedUserId),
                        userMitigationRequest);

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(200, response.getStatusCode());
        assertTrue(response.getBody().contains("success"));
        verify(mockPendingMitigationService)
                .persistPendingMitigation(userMitigationRequest, "456", "POST");
    }

    @Test
    public void shouldReturnBadRequestWhenInvalidEndpoint() throws IOException {
        APIGatewayProxyRequestEvent event = createTestEvent("POST", "/user/123/invalid", "{}");

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(400, response.getStatusCode());
        assertTrue(response.getBody().contains("Invalid URI."));
    }

    @Test
    public void shouldReturnBadRequestWhenInvalidRequestBodyForMitigation() throws IOException {
        APIGatewayProxyRequestEvent event =
                createTestEvent("POST", "/user/123/cis", "invalid json");

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(400, response.getStatusCode());
        assertTrue(response.getBody().contains("Invalid request body"));
    }

    @Test
    public void shouldReturnDataAlreadyExistForUserCis() throws IOException {
        UserCisRequest userCisRequest =
                UserCisRequest.builder()
                        .code("code1")
                        .issuanceDate("2023-07-25T10:00:00Z")
                        .mitigations(List.of("V01", "V03"))
                        .build();
        APIGatewayProxyRequestEvent event =
                createTestEvent("POST", "/user/123/cis", Collections.singletonList(userCisRequest));
        doThrow(new BadRequestException("User's CI Code cannot be null in all CIs"))
                .when(mockUserService)
                .addUserCis(anyString(), any());

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(400, response.getStatusCode());
        assertTrue(response.getBody().contains("User's CI Code cannot be null in all CIs"));
    }

    @Test
    public void shouldReturnDataNotFoundForUserCis() throws IOException {
        UserCisRequest userCisRequest =
                UserCisRequest.builder()
                        .code("code1")
                        .issuanceDate("2023-07-25T10:00:00Z")
                        .mitigations(List.of("V01", "V03"))
                        .build();
        APIGatewayProxyRequestEvent event =
                createTestEvent("PUT", "/user/123/cis", Collections.singletonList(userCisRequest));
        doThrow(new DataNotFoundException("User and ContraIndicator not found."))
                .when(mockUserService)
                .updateUserCis(anyString(), any());

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(404, response.getStatusCode());
        assertTrue(response.getBody().contains("User and ContraIndicator not found."));
    }

    @Test
    public void shouldReturnSuccessForValidPutRequestWithNoContent() throws IOException {
        APIGatewayProxyRequestEvent event = createTestEvent("PUT", "/user/123/cis", null);
        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(500, response.getStatusCode());
    }

    @Test
    public void shouldReturnSuccessForValidPutRequestWithEmptyContent() throws IOException {
        APIGatewayProxyRequestEvent event =
                createTestEvent("PUT", "/user/123/cis", Collections.emptyList());
        doNothing().when(mockUserService).updateUserCis(anyString(), anyList());

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(200, response.getStatusCode());
        verify(mockUserService).updateUserCis(eq("123"), eq(Collections.emptyList()));
    }

    @Test
    public void shouldReturnSuccessForValidPutRequestWithSingleContent() throws IOException {
        UserCisRequest userCisRequest =
                UserCisRequest.builder()
                        .code("code1")
                        .issuanceDate("2023-07-25T10:00:00Z")
                        .mitigations(List.of("V01", "V03"))
                        .build();

        APIGatewayProxyRequestEvent event =
                createTestEvent("PUT", "/user/123/cis", Collections.singletonList(userCisRequest));
        doNothing().when(mockUserService).updateUserCis(anyString(), anyList());

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(200, response.getStatusCode());
        verify(mockUserService)
                .updateUserCis(eq("123"), eq(Collections.singletonList(userCisRequest)));
    }

    @Test
    public void shouldReturnInternalServerErrorForUnknownException() throws IOException {
        UserCisRequest userCisRequest =
                UserCisRequest.builder()
                        .code("code1")
                        .issuanceDate("2023-07-25T10:00:00Z")
                        .mitigations(List.of("V01", "V03"))
                        .build();
        APIGatewayProxyRequestEvent event =
                createTestEvent("POST", "/user/123/cis", Collections.singletonList(userCisRequest));
        doThrow(new RuntimeException("Unknown exception occurred."))
                .when(mockUserService)
                .addUserCis(anyString(), anyList());

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(500, response.getStatusCode());
        assertTrue(response.getBody().contains("Unknown exception occurred."));
    }

    @Test
    public void shouldReturnBadRequestForInvalidMethod() throws IOException {
        UserCisRequest userCisRequest =
                UserCisRequest.builder()
                        .code("code1")
                        .issuanceDate("2023-07-25T10:00:00Z")
                        .mitigations(List.of("V01", "V03"))
                        .build();
        APIGatewayProxyRequestEvent event =
                createTestEvent(
                        "PATCH", "/user/123/cis", Collections.singletonList(userCisRequest));

        APIGatewayProxyResponseEvent response =
                stubManagementHandler.handleRequest(event, mock(Context.class));

        assertEquals(400, response.getStatusCode());
        assertTrue(response.getBody().contains("Http Method is not supported."));
    }

    private APIGatewayProxyRequestEvent createTestEvent(String httpMethod, String path, Object body)
            throws JsonProcessingException {
        Map<String, String> pathParameters = new HashMap<>();
        pathParameters.put("userId", "123");
        pathParameters.put("ci", "456");

        APIGatewayProxyRequestEvent event =
                new APIGatewayProxyRequestEvent().withHttpMethod(httpMethod).withPath(path);

        event.setPathParameters(pathParameters);
        if (body != null) {
            event.setBody(objectMapper.writeValueAsString(body));
        }

        return event;
    }
}

File: di-ipv-cimit-stub/lambdas/test/external-api/stub-management/stubManagementHandler.test.ts
import {
  APIGatewayEventDefaultAuthorizerContext,
  APIGatewayEventRequestContextWithAuthorizer,
} from "aws-lambda/common/api-gateway";

jest.mock(
  "../../../src/external-api/stub-management/service/userService",
  () => ({
    addUserCis: jest.fn(),
    updateUserCis: jest.fn(),
  }),
);

jest.mock("../../../src/common/cimitStubItemService", () => ({
  getCiForUserId: jest.fn(),
}));

jest.mock("../../../src/common/pendingMitigationService", () => ({
  persistPendingMitigation: jest.fn(),
}));

import { APIGatewayProxyEvent } from "aws-lambda";
import { handler } from "../../../src/external-api/stub-management/stubManagementHandler";
import * as userService from "../../../src/external-api/stub-management/service/userService";
import * as cimitStubItemService from "../../../src/common/cimitStubItemService";
import * as pendingMitigationService from "../../../src/common/pendingMitigationService";

describe("stubManagementHandler", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const createEvent = (
    httpMethod: string,
    path: string,
    pathParameters: { [key: string]: string },
    body?: object,
  ): APIGatewayProxyEvent => ({
    httpMethod,
    path,
    body: body !== undefined ? JSON.stringify(body) : null,
    pathParameters: pathParameters,
    headers: {},
    multiValueHeaders: {},
    isBase64Encoded: false,
    queryStringParameters: null,
    multiValueQueryStringParameters: null,
    stageVariables: null,
    requestContext:
      {} as APIGatewayEventRequestContextWithAuthorizer<APIGatewayEventDefaultAuthorizerContext>,
    resource: "",
  });

  describe("CI requests", () => {
    it("should add user CIs successfully when valid CI request list", async () => {
      const userCisRequests = [
        {
          code: "code1",
          issuanceDate: "2023-07-25T10:00:00Z",
          mitigations: ["V01", "V03"],
          issuer: "",
          document: "",
          txn: "",
        },
        {
          code: "code2",
          issuanceDate: "2023-07-25T10:00:00Z",
          mitigations: [],
          issuer: "",
          document: "",
          txn: "",
        },
      ];

      const event = createEvent(
        "POST",
        "/user/123/cis",
        { userId: "123" },
        userCisRequests,
      );
      const response = await handler(event);

      expect(response.statusCode).toBe(200);
      expect(response.body).toContain("success");
      expect(userService.addUserCis).toHaveBeenCalledWith(
        "123",
        userCisRequests,
      );
    });

    it("should update user Cis from valid CI request", async () => {
      const userCisRequest = [
        {
          code: "code1",
          issuanceDate: "2023-07-25T10:00:00Z",
          mitigations: ["V01", "V03"],
          issuer: "",
          document: "",
          txn: "",
        },
      ];

      const event = createEvent(
        "PUT",
        "/user/123/cis",
        { userId: "123" },
        userCisRequest,
      );
      const response = await handler(event);

      expect(response.statusCode).toBe(200);
      expect(response.body).toContain("success");
      expect(userService.updateUserCis).toHaveBeenCalledWith(
        "123",
        userCisRequest,
      );
    });

    it("should handle default user ID format in CI endpoint pattern", async () => {
      const urlEncodedUserId =
        "urn%3Auuid%3Ac08630f8-330e-43f8-a782-21432a197fc5";
      const userCisRequests = [
        {
          code: "code1",
          issuanceDate: "2023-07-25T10:00:00Z",
          mitigations: ["V01", "V03"],
          issuer: "",
          document: "",
          txn: "",
        },
      ];

      const event = createEvent(
        "POST",
        `/user/${urlEncodedUserId}/cis`,
        { userId: urlEncodedUserId },
        userCisRequests,
      );
      const response = await handler(event);

      expect(response.statusCode).toBe(200);
      expect(response.body).toContain("success");
      expect(userService.addUserCis).toHaveBeenCalledWith(
        "urn:uuid:c08630f8-330e-43f8-a782-21432a197fc5",
        userCisRequests,
      );
    });

    it("should return bad request when invalid request body for CIs", async () => {
      const event = createEvent("POST", "/user/123/cis", { userId: "123" });
      event.body = "invalid json";

      const response = await handler(event);

      expect(response.statusCode).toBe(400);
      expect(response.body).toContain("Invalid request body");
    });

    it("should return bad request for empty code", async () => {
      const userCisRequest = [
        {
          code: "",
          issuanceDate: "2023-07-25T10:00:00Z",
          mitigations: ["V01", "V03"],
          issuer: "",
          document: "",
          txn: "",
        },
      ];

      const event = createEvent(
        "POST",
        "/user/123/cis",
        { userId: "123" },
        userCisRequest,
      );
      const response = await handler(event);

      expect(response.statusCode).toBe(400);
      expect(response.body).toContain("CI codes cannot be empty.");
      expect(userService.addUserCis).not.toHaveBeenCalled();
    });

    it("should return success for valid PUT request with empty content", async () => {
      const event = createEvent("PUT", "/user/123/cis", { userId: "123" }, []);
      const response = await handler(event);

      expect(response.statusCode).toBe(200);
      expect(userService.updateUserCis).toHaveBeenCalledWith("123", []);
    });

    it("should return success for valid PUT request with single content", async () => {
      const userCisRequest = [
        {
          code: "code1",
          issuanceDate: "2023-07-25T10:00:00Z",
          mitigations: ["V01", "V03"],
          issuer: "",
          document: "",
          txn: "",
        },
      ];

      const event = createEvent(
        "PUT",
        "/user/123/cis",
        { userId: "123" },
        userCisRequest,
      );
      const response = await handler(event);

      expect(response.statusCode).toBe(200);
      expect(userService.updateUserCis).toHaveBeenCalledWith(
        "123",
        userCisRequest,
      );
    });
  });

  describe("Mitigation requests", () => {
    ["POST", "PUT"].forEach((method) => {
      it(`should add pending mitigation when valid mitigation request - ${method}`, async () => {
        (
          cimitStubItemService.getCiForUserId as jest.Mock
        ).mockResolvedValueOnce([{}]);
        const userMitigationRequest = { mitigations: ["V01"], vcJti: "jti123" };

        const event = createEvent(
          method,
          "/user/123/mitigations/456",
          { userId: "123", ci: "456" },
          userMitigationRequest,
        );
        const response = await handler(event);

        expect(response.statusCode).toBe(200);
        expect(response.body).toContain("success");
        expect(cimitStubItemService.getCiForUserId).toHaveBeenCalledWith(
          "123",
          "456",
        );
        expect(
          pendingMitigationService.persistPendingMitigation,
        ).toHaveBeenCalledWith(userMitigationRequest, "456", method);
      });
    });

    ["POST", "PUT"].forEach((method) => {
      it(`should return 404 if CI item not found for mitigation request - ${method}`, async () => {
        (
          cimitStubItemService.getCiForUserId as jest.Mock
        ).mockResolvedValueOnce(null);
        const userMitigationRequest = { mitigations: ["V01"], vcJti: "jti123" };

        const event = createEvent(
          method,
          "/user/123/mitigations/456",
          { userId: "123", ci: "456" },
          userMitigationRequest,
        );
        const response = await handler(event);

        expect(response.statusCode).toBe(404);
        expect(response.body).toContain("not found");
        expect(cimitStubItemService.getCiForUserId).toHaveBeenCalledWith(
          "123",
          "456",
        );
        expect(
          pendingMitigationService.persistPendingMitigation,
        ).not.toHaveBeenCalled();
      });
    });

    ["POST", "PUT"].forEach((method) => {
      it("should return bad request when invalid request body for mitigation", async () => {
        const event = createEvent(method, "/user/123/mitigations/456", {
          userId: "123",
          ci: "456",
        });
        event.body = "invalid json";

        const response = await handler(event);

        expect(response.statusCode).toBe(400);
        expect(response.body).toContain("Invalid request body");
      });
    });

    ["POST", "PUT"].forEach((method) => {
      it("should handle default user ID format in mitigations pattern", async () => {
        (
          cimitStubItemService.getCiForUserId as jest.Mock
        ).mockResolvedValueOnce([{}]);
        const urlEncodedUserId =
          "urn%3Auuid%3Ac08630f8-330e-43f8-a782-21432a197fc5";
        const userMitigationRequest = { mitigations: ["V01"], vcJti: "jti123" };

        const event = createEvent(
          method,
          `/user/${urlEncodedUserId}/mitigations/456`,
          { userId: urlEncodedUserId, ci: "456" },
          userMitigationRequest,
        );
        const response = await handler(event);

        expect(response.statusCode).toBe(200);
        expect(response.body).toContain("success");
        expect(cimitStubItemService.getCiForUserId).toHaveBeenCalledWith(
          "urn:uuid:c08630f8-330e-43f8-a782-21432a197fc5",
          "456",
        );
        expect(
          pendingMitigationService.persistPendingMitigation,
        ).toHaveBeenCalledWith(userMitigationRequest, "456", method);
      });
    });

    ["POST", "PUT"].forEach((method) => {
      it("should return data not found for missing user CIs", async () => {
        const userCisRequest = [
          {
            code: "code1",
            issuanceDate: "2023-07-25T10:00:00Z",
            mitigations: ["V01", "V03"],
            issuer: "",
            document: "",
            txn: "",
          },
        ];
        (
          cimitStubItemService.getCiForUserId as jest.Mock
        ).mockResolvedValueOnce([]);

        const event = createEvent(
          method,
          "/user/123/mitigations/456",
          { userId: "123", ci: "456" },
          userCisRequest,
        );
        const response = await handler(event);

        expect(response.statusCode).toBe(404);
        expect(response.body).toContain("User and ContraIndicator not found.");
        expect(cimitStubItemService.getCiForUserId).toHaveBeenCalledWith(
          "123",
          "456",
        );
      });
    });
  });

  it("should return bad request when invalid endpoint", async () => {
    const event = createEvent("POST", "/user/123/invalid", {});
    const response = await handler(event);

    expect(response.statusCode).toBe(400);
    expect(response.body).toContain("Invalid URI.");
  });

  it("should return error for CI request with no content", async () => {
    const event = createEvent("PUT", "/user/123/cis", { userId: "123" });
    event.body = null;

    const response = await handler(event);

    expect(response.statusCode).toBe(400);
    expect(response.body).toContain("Invalid request body");
  });

  it("should return internal server error for unknown exception", async () => {
    const userCisRequest = [
      {
        code: "code1",
        issuanceDate: "2023-07-25T10:00:00Z",
        mitigations: ["V01", "V03"],
        issuer: "",
        document: "",
        txn: "",
      },
    ];
    (userService.addUserCis as jest.Mock).mockRejectedValueOnce(
      new Error("Unknown exception occurred."),
    );

    const event = createEvent(
      "POST",
      "/user/123/cis",
      { userId: "123" },
      userCisRequest,
    );
    const response = await handler(event);

    expect(response.statusCode).toBe(500);
    expect(response.body).toContain("Unknown exception occurred.");
  });

  it("should return bad request for invalid method", async () => {
    const userCisRequest = [
      {
        code: "code1",
        issuanceDate: "2023-07-25T10:00:00Z",
        mitigations: ["V01", "V03"],
        issuer: "",
        document: "",
        txn: "",
      },
    ];
    const event = createEvent(
      "PATCH",
      "/user/123/cis",
      { userId: "123" },
      userCisRequest,
    );

    const response = await handler(event);

    expect(response.statusCode).toBe(400);
    expect(response.body).toContain("Http Method is not supported.");
  });
});
